### ISSUE TITLE
Unable to authenticate with az login

### ISSUE THREAD

**Issue by a01fe** [ansible/ansible_collections_azure#765]
##### SUMMARY

After authenticating with `az login`, azure modules fail with `Failed to authenticate with provided credentials`.

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

authentication

##### ANSIBLE VERSION

```paste below
ansible [core 2.12.2]
  config file = /home/xxx/Projects/aks/ansible.cfg
  configured module search path = ['/home/xxx/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /home/xxx/Projects/aks/.venv/lib/python3.9/site-packages/ansible
  ansible collection location = /home/xxx/Projects/aks
  executable location = /home/xxx/Projects/aks/.venv/bin/ansible
  python version = 3.9.10 (main, Jan 15 2022, 18:17:56) [GCC 9.3.0]
  jinja version = 3.0.3
  libyaml = True
```

##### COLLECTION VERSION

```paste below
# /home/xxx/Projects/aks/.venv/lib/python3.9/site-packages/ansible_collections
Collection         Version
------------------ -------
azure.azcollection 1.11.0 

# /home/xxx/Projects/aks/ansible_collections
Collection         Version
------------------ -------
azure.azcollection 1.11.0 
```

##### CONFIGURATION

```paste below
ANSIBLE_NOCOWS(/home/xxx/Projects/aks/ansible.cfg) = True
COLLECTIONS_PATHS(/home/xxx/Projects/aks/ansible.cfg) = ['/home/xxx/Projects/aks']
DEFAULT_ROLES_PATH(/home/xxx/Projects/aks/ansible.cfg) = ['/home/xxx/Projects/aks/roles']
DEFAULT_STDOUT_CALLBACK(/home/xxx/Projects/aks/ansible.cfg) = yaml
```

##### OS / ENVIRONMENT

Windows, WSL2 Ubuntu 20.04

##### STEPS TO REPRODUCE

Run:

```bash
az login
ansible-playbook  test_playbook.yaml
```

```yaml
- name: Test azure authentication
  gather_facts: false
  hosts: localhost
  connection: local
  tasks:
    - name: Get resource groups
      azure.azcollection.azure_rm_resourcegroup_info:
      register: rg_info

    - name: show resource groups
      debug:
        var: rg_info
```

##### EXPECTED RESULTS

No failures and a list of resource groups.

##### ACTUAL RESULTS

`Get resource groups` task fails.

```paste below
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [Test azure authentication] **********************************************************************************************************************************************************

TASK [Get resource groups] ****************************************************************************************************************************************************************
fatal: [localhost]: FAILED! => changed=false 
  msg: Failed to authenticate with provided credentials. Some attributes were missing. Credentials must include client_id, secret and tenant or ad_user and password, or ad_user, password, client_id, tenant and adfs_authority_url(optional) for ADFS authentication, or be logged in using AzureCLI.

PLAY RECAP ********************************************************************************************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0   
```
##### Additional info

Output from `az --version`:

```
azure-cli                         2.33.1

core                              2.33.1
telemetry                          1.0.6

Dependencies:
msal                              1.16.0
azure-mgmt-resource               20.0.0

Python location '/opt/az/bin/python3'
Extensions directory '/home/xxx/.azure/cliextensions'

Python (Linux) 3.6.10 (default, Feb 14 2022, 04:04:37) 
[GCC 9.3.0]

Legal docs and information: aka.ms/AzureCliLegal


Your CLI is up-to-date.

Please let us know how we are doing: https://aka.ms/azureclihats
and let us know if you're interested in trying out our newest features: https://aka.ms/CLIUXstudy
```

After running `az login`, `~/.azure/accessTokens.json` does not exist.
This Microsoft page says `accessTokens.json` is deprecated.

https://docs.microsoft.com/en-us/cli/azure/msal-based-azure-cli

How can we authenticate without SPs? Our AD admins do not want to create SPs for every developer.

**Comment by johnpetersjr** [comment]
Have you tried putting this in your main playbook?
```
  vars:
    client_id: "{{lookup('env', 'AZURE_CLIENT_ID')}}"
    secret: "{{lookup('env', 'AZURE_SECRET')}}"
    subscription_id: "{{lookup('env', 'AZURE_SUBSCRIPTION_ID')}}"
    tenant: "{{lookup('env', 'AZURE_TENANT')}}"
```
I need to do that in all my azure / azcollection plays...

What auth_source are you trying to use?

**Comment by a01fe** [comment]
We're using `cli` as the auth source. It works for me, but it fails for new installs with the current version of az cli.

I've tried creating an sp with `az ad sp create-for-rbac`, setting `AZURE_CLIENT_ID`, `AZURE_SECRET`, `AZURE_TENANT` from the json it returns, along with `AZURE_SUBSCRIPTION_ID` and `AZURE_CLOUD_ENVIRONMENT`.

That gets a different error:

```
The subscription 'xxx-xxx-xxx-xxx-xxx' could not be found
```

The subscription in the error message is correct, it's not clear why that isn't working.

**Comment by Fred-sun** [comment]
@a01fe Did you check to see if your sub was correct? You can try "az login" and "az account set -s "subscription_id". Thank you very much!

**Comment by a01fe** [comment]
After doing `az account set -s <my-sub-id>`, now I get a traceback:

```
[WARNING]: No inventory was parsed, only implicit localhost is available
[WARNING]: provided hosts list is empty, only localhost is available. Note that the implicit localhost does not match 'all'

PLAY [Test azure authentication] ************************************************************************************************
TASK [Get resource groups] ******************************************************************************************************
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: knack.util.CLIError: Could not retrieve token from local cache. Please run 'az login'.
fatal: [localhost]: FAILED! => changed=false
  module_stderr: |-
    Traceback (most recent call last):
      File "/home/xxx/.ansible/tmp/ansible-tmp-1645544288.4691625-499-210961925680216/AnsiballZ_azure_rm_resourcegroup_info.py", line 107, in <module>
        _ansiballz_main()
      File "/home/xxx/.ansible/tmp/ansible-tmp-1645544288.4691625-499-210961925680216/AnsiballZ_azure_rm_resourcegroup_info.py", line 99, in _ansiballz_main
        invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
      File "/home/xxx/.ansible/tmp/ansible-tmp-1645544288.4691625-499-210961925680216/AnsiballZ_azure_rm_resourcegroup_info.py", line 47, in invoke_module
        runpy.run_module(mod_name='ansible_collections.azure.azcollection.plugins.modules.azure_rm_resourcegroup_info', init_globals=dict(_module_fqn='ansible_collections.azure.azcollection.plugins.modules.azure_rm_resourcegroup_info', _modlib_path=modlib_path),
      File "/usr/lib/python3.9/runpy.py", line 210, in run_module
        return _run_module_code(code, init_globals, run_name, mod_spec)
      File "/usr/lib/python3.9/runpy.py", line 97, in _run_module_code
        _run_code(code, mod_globals, init_globals,
      File "/usr/lib/python3.9/runpy.py", line 87, in _run_code
        exec(code, run_globals)
      File "/tmp/ansible_azure.azcollection.azure_rm_resourcegroup_info_payload_4d8dw4r8/ansible_azure.azcollection.azure_rm_resourcegroup_info_payload.zip/ansible_collections/azure/azcollection/plugins/modules/azure_rm_resourcegroup_info.py", line 233, in <module>
      File "/tmp/ansible_azure.azcollection.azure_rm_resourcegroup_info_payload_4d8dw4r8/ansible_azure.azcollection.azure_rm_resourcegroup_info_payload.zip/ansible_collections/azure/azcollection/plugins/modules/azure_rm_resourcegroup_info.py", line 229, in main
      File "/tmp/ansible_azure.azcollection.azure_rm_resourcegroup_info_payload_4d8dw4r8/ansible_azure.azcollection.azure_rm_resourcegroup_info_payload.zip/ansible_collections/azure/azcollection/plugins/modules/azure_rm_resourcegroup_info.py", line 158, in __init__
      File "/tmp/ansible_azure.azcollection.azure_rm_resourcegroup_info_payload_4d8dw4r8/ansible_azure.azcollection.azure_rm_resourcegroup_info_payload.zip/ansible_collections/azure/azcollection/plugins/module_utils/azure_rm_common.py", line 471, in __init__
      File "/tmp/ansible_azure.azcollection.azure_rm_resourcegroup_info_payload_4d8dw4r8/ansible_azure.azcollection.azure_rm_resourcegroup_info_payload.zip/ansible_collections/azure/azcollection/plugins/modules/azure_rm_resourcegroup_info.py", line 174, in exec_module
      File "/tmp/ansible_azure.azcollection.azure_rm_resourcegroup_info_payload_4d8dw4r8/ansible_azure.azcollection.azure_rm_resourcegroup_info_payload.zip/ansible_collections/azure/azcollection/plugins/modules/azure_rm_resourcegroup_info.py", line 209, in list_items
      File "/home/xxx/Projects/aks/.venv/lib/python3.9/site-packages/msrest/paging.py", line 143, in __next__
        self.advance_page()
      File "/home/xxx/Projects/aks/.venv/lib/python3.9/site-packages/msrest/paging.py", line 129, in advance_page
        self._response = self._get_next(self.next_link)
      File "/home/xxx/Projects/aks/.venv/lib/python3.9/site-packages/azure/mgmt/resource/resources/v2017_05_10/operations/_resource_groups_operations.py", line 510, in internal_paging
        response = self._client.send(request, stream=False, **operation_config)
      File "/home/xxx/Projects/aks/.venv/lib/python3.9/site-packages/msrest/service_client.py", line 336, in send
        pipeline_response = self.config.pipeline.run(request, **kwargs)
      File "/home/xxx/Projects/aks/.venv/lib/python3.9/site-packages/msrest/pipeline/__init__.py", line 197, in run
        return first_node.send(pipeline_request, **kwargs)  # type: ignore
      File "/home/xxx/Projects/aks/.venv/lib/python3.9/site-packages/msrest/pipeline/__init__.py", line 150, in send
        response = self.next.send(request, **kwargs)
      File "/home/xxx/Projects/aks/.venv/lib/python3.9/site-packages/msrest/pipeline/requests.py", line 65, in send
        self._creds.signed_session(session)
      File "/home/xxx/Projects/aks/.venv/lib/python3.9/site-packages/azure/cli/core/adal_authentication.py", line 125, in signed_session
        scheme, token, _, external_tenant_tokens = self._get_token()
      File "/home/xxx/Projects/aks/.venv/lib/python3.9/site-packages/azure/cli/core/adal_authentication.py", line 44, in _get_token
        raise err
      File "/home/xxxx/Projects/aks/.venv/lib/python3.9/site-packages/azure/cli/core/adal_authentication.py", line 38, in _get_token
        scheme, token, token_entry = self._token_retriever(sdk_resource)
      File "/home/xxx/Projects/aks/.venv/lib/python3.9/site-packages/azure/cli/core/_profile.py", line 587, in _retrieve_token
        return self._creds_cache.retrieve_token_for_user(username_or_sp_id,
      File "/home/xxx/Projects/aks/.venv/lib/python3.9/site-packages/azure/cli/core/_profile.py", line 1106, in retrieve_token_for_user
        raise CLIError("Could not retrieve token from local cache.{}".format(
    knack.util.CLIError: Could not retrieve token from local cache. Please run 'az login'.
  module_stdout: ''
  msg: |-
    MODULE FAILURE
    See stdout/stderr for the exact error
  rc: 1

PLAY RECAP **********************************************************************************************************************
localhost                  : ok=0    changed=0    unreachable=0    failed=1    skipped=0    rescued=0    ignored=0
```

Thanks for your help!

**Comment by Fred-sun** [comment]
@a01fe This should be your account problem, can not get access permission, you can refer to the following two links to try! Thank you very much!
```e.g.
https://www.teosoft.it/post/2018-02-28-personal-vpn-with-algo-on-azure/
https://docs.microsoft.com/en-us/azure/active-directory/develop/howto-create-service-principal-portal

**Comment by a01fe** [comment]
Are you saying that az cli authentication will no longer be supported?

I can `az login` on my mac and successfully run `test_playbook.yaml`, as can all my teammates except for a new team member using a new az cli install.

If I copy `~/.azure/accessTokens.json` from my mac to a workstation where the playbook is failing, then the playbook works.

**Comment by verhyppo** [comment]
Hello, 

we encountered a similar problem with the latest version of the az-cli.
Basically we found out that az-cli version needs to match the version of az-cli-core dependency.

https://github.com/ansible-collections/azure/blob/765e82d5021930bff8b45cf7c6edded79d7e6ef6/requirements-azure.txt#L4

Let's have a look at the output of the command
```
az --version
```
to find out which version of the az-cli (and core)  you are using.
I experienced this problem mainly on Ubuntu, don't know about other OSes.
